### PR TITLE
First look: Upgrade to Phoenix 1.3

### PR DESCRIPTION
**Pull request by andrewhao**
razor

Overview
---

This is a manual migration to Phoenix 1.3, following a mix of approaches from the [upgrade guide](http://phoenixframework.org/blog/upgrading-from-120-to-130) to using the actual diff from [Phoenixdiff](https://www.phoenixdiff.org/?source=1.2.5&target=1.3.0) 

This addresses issue https://github.com/carbonfive/razor-phoenix/issues/21.

Changelog
---
1. Migrated to Phoenix 1.3 with new contexts
2. Runtime configs for Endpoint and Repo configs are now (mostly) done via environment variables in `Repo.init/2` and `Endpoint.init/2`.
3. Upgraded Erlang to 20.1, Elixir to 1.5.2, Node to 8.9.3.
4. Introduced a `phoenix_static_buildpack.config` file to manage Node configurations.
5. MIX_ENV is now standardized to `prod`; `production.ex` renamed to `prod.ex`.
6. Removed the separate `acceptance.ex` configuration. Now all Heroku deployments should have `MIX_ENV=prod` and be configured via environment variables.
7. Upgraded Ecto, Wallaby, ExMachina and other related libraries to latest.

Steps
---
- [x] Move files, configs around
- [x] Make tests pass
- [x] Test with Razor cutting and custom app name setup (https://github.com/andrewhao/toystore)
- [x] File/configuration parity with a fresh 1.3 application (full diff resolution)
- [x] Test cut application deployment to Heroku: http://sample-phoenix-toy-store.herokuapp.com/

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -7,7 +7,6 @@ defmodule AppPrototype.Mixfile do
      elixir: "~> 1.5.1",
      elixirc_paths: elixirc_paths(Mix.env),
      compilers: [:phoenix, :gettext] ++ Mix.compilers,
-     build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
```
Comment (andrewhao) [inline mix.exs]: @craiglyons This might be an issue in a different PR, but if our convention is now `MIX_ENV=production`, then should this be `Mix.env == :production`?

File: mix.exs [hunk 2]
```diff
@@ -20,31 +19,32 @@ defmodule AppPrototype.Mixfile do
   #
   # Type `mix help compile.app` for more information.
   def application do
-    [mod: {AppPrototype, []},
-     applications: [:phoenix, :phoenix_pubsub, :phoenix_html, :cowboy, :logger, :gettext,
-                    :phoenix_ecto, :postgrex]]
+    [mod: {AppPrototype.Application, []},
```
Comment (andrewhao) [inline mix.exs]: Note that application definition moves into its own `Application` - see `application.ex`.

File: lib/app_prototype/repo.ex [hunk 3]
```diff
@@ -1,3 +1,11 @@
 defmodule AppPrototype.Repo do
   use Ecto.Repo, otp_app: :app_prototype
+
+  @doc """
+  Dynamically loads the repository url from the
+  DATABASE_URL environment variable.
+  """
+  def init(_, opts) do
+    {:ok, Keyword.put(opts, :url, System.get_env("DATABASE_URL"))}
```
Comment (andrewhao) [inline lib/app_prototype/repo.ex]: This is new Phoenix convention, allowing us to pick up the new config at runtime and avoiding compile-time issues.

File: lib/app_prototype_web/endpoint.ex [hunk 4]
```diff
@@ -43,5 +43,20 @@ defmodule AppPrototype.Endpoint do
...
+  
+  @doc """
+  Callback invoked for dynamically configuring the endpoint.
+
+  It receives the endpoint configuration and checks if
+  configuration should be loaded from the system environment.
+  """
+  def init(_key, config) do
+    if config[:load_from_system_env] do
+      port = System.get_env("PORT") || raise "expected the PORT environment variable to be set"
```
Comment (andrewhao) [inline lib/app_prototype_web/endpoint.ex]: Runtime config.

File: config/production.exs [hunk 5]
```diff
@@ -1,15 +1,68 @@
...
+# Don't forget to configure the url host to something meaningful,
+# Phoenix uses this information when generating URLs.
+#
+# Finally, we also include the path to a cache manifest
+# containing the digested version of static files. This
+# manifest is generated by the mix phx.digest task
+# which you typically run after static files are built.
+
+config :app_prototype, AppPrototypeWeb.Endpoint,
+  load_from_system_env: true
```
Comment (andrewhao) [inline config/production.exs]: Note the new `load_from_system_env` flag.

File: config/production.exs [hunk 6]
```diff
@@ -1,15 +1,68 @@
...
+  url: [scheme: "https", host: "app-prototype-prod.herokuapp.com", port: 443],
+  force_ssl: [rewrite_on: [:x_forwarded_proto]],
+  cache_static_manifest: "priv/static/cache_manifest.json",
+  secret_key_base: System.get_env("SECRET_KEY_BASE")
+
+# ## SSL Support
```
Comment (andrewhao) [inline config/production.exs]: I threw this comment block in just because the generator provides it - did we omit it previously on purpose @craiglyons?

File: mix.exs (hunk 1 above)
Comment (craiglyons) [reply mix.exs]: For posterity, we agreed to normalize on `prod`. 👍

**Comment by andrewhao** [comment]
@craiglyons Done. I'll merge when the build passes. Thanks! 🌟
